sql/core/src/test/resources/sql-tests/results/udf/udf-group-by.sql.out

```diff
@@ -0,0 +1,521 @@
+struct<>
+-- !query 1 output
+org.apache.spark.sql.AnalysisException
+grouping expressions sequence is empty, and 'testdata.`a`' is not an aggregate function. Wrap '(CAST(udf(cast(count(b) as string)) AS BIGINT) AS `CAST(udf(cast(count(b) as string)) AS BIGINT)`)' in windowing function(s) or wrap 'testdata.`a`' in first() (or first_value) if you don't care which value you get.;
+
+
+-- !query 2
+SELECT COUNT(udf(a)), udf(COUNT(b)) FROM testData
+-- !query 2 schema
+struct<count(CAST(udf(cast(a as string)) AS INT)):bigint,CAST(udf(cast(count(b) as string)) AS BIGINT):bigint>
+-- !query 2 output
+7	7
+
+
+-- !query 3
+SELECT udf(a), COUNT(udf(b)) FROM testData GROUP BY a
+-- !query 3 schema
+struct<CAST(udf(cast(a as string)) AS INT):int,count(CAST(udf(cast(b as string)) AS INT)):bigint>
+-- !query 3 output
+1	2
+2	2
+3	2
+NULL	1
+
+
+-- !query 4
+SELECT udf(a), udf(COUNT(udf(b))) FROM testData GROUP BY b
+-- !query 4 schema
+struct<>
+-- !query 4 output
+org.apache.spark.sql.AnalysisException
+expression 'testdata.`a`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;
+
+
+-- !query 5
+SELECT COUNT(udf(a)), COUNT(udf(b)) FROM testData GROUP BY udf(a)
+-- !query 5 schema
+struct<count(CAST(udf(cast(a as string)) AS INT)):bigint,count(CAST(udf(cast(b as string)) AS INT)):bigint>
+-- !query 5 output
+0	1
+2	2
+2	2
+3	2
+
+
+-- !query 6
+SELECT 'foo', COUNT(udf(a)) FROM testData GROUP BY 1
+-- !query 6 schema
+struct<foo:string,count(CAST(udf(cast(a as string)) AS INT)):bigint>
+-- !query 6 output
+foo	7
+
+
+-- !query 7
+SELECT 'foo' FROM testData WHERE a = 0 GROUP BY udf(1)
+-- !query 7 schema
+struct<foo:string>
+-- !query 7 output
+
+
+
+-- !query 8
+SELECT 'foo', udf(APPROX_COUNT_DISTINCT(udf(a))) FROM testData WHERE a = 0 GROUP BY 1
+-- !query 8 schema
+struct<foo:string,CAST(udf(cast(approx_count_distinct(cast(udf(cast(a as string)) as int), 0.05, 0, 0) as string)) AS BIGINT):bigint>
+-- !query 8 output
+
+
+
+-- !query 9
+SELECT 'foo', MAX(STRUCT(udf(a))) FROM testData WHERE a = 0 GROUP BY 1
+-- !query 9 schema
+struct<foo:string,max(named_struct(col1, CAST(udf(cast(a as string)) AS INT))):struct<col1:int>>
+-- !query 9 output
+
+
+
+-- !query 10
+SELECT udf(a + b), udf(COUNT(b)) FROM testData GROUP BY a + b
+-- !query 10 schema
+struct<CAST(udf(cast((a + b) as string)) AS INT):int,CAST(udf(cast(count(b) as string)) AS BIGINT):bigint>
+-- !query 10 output
+2	1
+3	2
+4	2
+5	1
+NULL	1
+
+
+-- !query 11
+SELECT udf(a + 2), udf(COUNT(b)) FROM testData GROUP BY a + 1
+-- !query 11 schema
+struct<>
+-- !query 11 output
+org.apache.spark.sql.AnalysisException
+expression 'testdata.`a`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;
+
+
+-- !query 12
+SELECT udf(a + 1 + 1), udf(COUNT(b)) FROM testData GROUP BY udf(a + 1)
```

Comment:
Following original group-by.sql, I think this should be `SELECT udf(a + 1) + 1, udf(COUNT(b)) FROM testData GROUP BY udf(a + 1)`?

Reply:
Yeah it could be both should work though.

Reply:
Since this query is grouping by `udf(a + 1)`, `udf(a + 1 + 1)` is an expression the analyzer will complain about. (an AnalysisException)

Reply:
yes you are right.. its based on group value. Will change it.

Reply:
Actually @viirya it still fails:
```
-- !query 12
SELECT udf(a + 1) + 1, udf(COUNT(b)) FROM testData GROUP BY udf(a + 1)
-- !query 12 schema
struct<>
-- !query 12 output
org.apache.spark.sql.AnalysisException
expression 'testdata.`a`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;
```

Reply:
@viirya @HyukjinKwon  If I do that scala udf will generate:
```
 SELECT udf(a + 1) + 1, udf(COUNT(b)) FROM testData GROUP BY udf(a + 1)
 -- !query 12 schema
-struct<>
+struct<(CAST(udf(cast((a + 1) as string)) AS INT) + 1):int,CAST(udf(cast(count(b) as string)) AS BIGINT):bigint>
 -- !query 12 output
-org.apache.spark.sql.AnalysisException
-expression 'testdata.`a`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;
+3      2
+4      2
+5      2
+NULL   1
```
bu then python and pandas udfs will fail with that exception.

Reply:
here is the output:
```
Using Python version 2.7.15+ (default, Nov 27 2018 23:36:35)
SparkSession available as 'spark'.
>>> from pyspark.sql.functions import pandas_udf, PandasUDFType
>>> @pandas_udf("string", PandasUDFType.SCALAR)
... def noop(x):
...     return x.apply(str)
... 

>>> spark.udf.register("udf", noop)
<function noop at 0x7ff7ea465b90>
>>> spark.sql("CREATE OR REPLACE TEMPORARY VIEW testData AS SELECT * FROM VALUES (1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2), (null, 1), (3, null), (null, null) AS testData(a, b)")

DataFrame[]
>>> 
>>> spark.sql("SELECT udf(a + 1) + 1, udf(COUNT(b)) FROM testData GROUP BY udf(a + 1)")
19/07/18 18:41:12 WARN ObjectStore: Failed to get database global_temp, returning NoSuchObjectException
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/pyspark/sql/session.py", line 806, in sql
    return DataFrame(self._jsparkSession.sql(sqlQuery), self._wrapped)
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/lib/py4j-0.10.8.1-src.zip/py4j/java_gateway.py", line 1286, in __call__
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/pyspark/sql/utils.py", line 93, in deco
    raise converted
pyspark.sql.utils.AnalysisException: u"expression 'testdata.`a`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;;\nAggregate [udf((a#0 + 1))], [(cast(udf((a#0 + 1)) as double) + cast(1 as double)) AS (CAST(udf((a + 1)) AS DOUBLE) + CAST(1 AS DOUBLE))#6, udf(count(b#1)) AS udf(count(b))#7]\n+- SubqueryAlias `testdata`\n   +- Project [a#0, b#1]\n      +- SubqueryAlias `testData`\n         +- LocalRelation [a#0, b#1]\n"
```

Reply:
Let's comment the test with some comments that explains this.

Reply:
ok will do.

Reply:
I took a look again. Actually, 

```sql
SELECT udf(a + 1 + 1), udf(COUNT(b)) FROM testData GROUP BY udf(a + 1)
```

is consistently being failed at `udf(a + 1)` because here seems expecting `udf` to be an aggregate function. It should be, for instance, Pandas group aggregate function IIRC; however, integrated UDF test base doesn't have it.

Reply:
Yes, let's skip, file a JIRA with some reproducers. I also tested by myself. feel free to copy and paste this in the new JIRA:

Python:

```python
from pyspark.sql.functions import pandas_udf, PandasUDFType

@pandas_udf("int", PandasUDFType.SCALAR)
def noop(x):
    return x

spark.udf.register("udf", noop)

sql("""
  CREATE OR REPLACE TEMPORARY VIEW testData AS SELECT * FROM VALUES
  (1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2), (null, 1), (3, null), (null, null)
  AS testData(a, b)""")

sql("""SELECT udf(a + 1), udf(COUNT(b)) FROM testData GROUP BY udf(a + 1)""").show()
```

```
: org.apache.spark.sql.AnalysisException: expression 'testdata.`a`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;;
Aggregate [udf((a#0 + 1))], [udf((a#0 + 1)) AS udf((a + 1))#10, udf(count(b#1)) AS udf(count(b))#12]
+- SubqueryAlias `testdata`
   +- Project [a#0, b#1]
      +- SubqueryAlias `testData`
         +- LocalRelation [a#0, b#1]
```

Scala:

```scala
spark.udf.register("udf", (input: Int) => input)

sql("""
  CREATE OR REPLACE TEMPORARY VIEW testData AS SELECT * FROM VALUES
  (1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2), (null, 1), (3, null), (null, null)
  AS testData(a, b)""")

sql("""SELECT udf(a + 1), udf(COUNT(b)) FROM testData GROUP BY udf(a + 1)""").show()
```

```
+------------+-------------+
|udf((a + 1))|udf(count(b))|
+------------+-------------+
|        null|            1|
|           3|            2|
|           4|            2|
|           2|            2|
+------------+-------------+
```

Let's skip the tests here with some comments that links the new JIRA, and merge this one first.

Reply:
Here's one example when I met such case.

Reply:
Cool I will create the jira thanks.

Reply:
oops, I mean here is the example: https://github.com/apache/spark/blob/master/sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-aggregates_part1.sql#L65-L67

Reply:
@HyukjinKwon [Jira](https://issues.apache.org/jira/browse/SPARK-28445), also added a pointer to it in the comment.

Reply:
Submitted #25215 to fix it.